Use Linux Sysfs SoC interface for SystemInfo and CPUInfo

superna9999 opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
Newer platforms upstreamed uses the newly introduced SoC sysfs interface to give userspace information about the system SoC family, id and revision. It also gives the current board model.

The interface is documented at : https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-devices-soc

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
Under Upstream Linux, the SystemInfo and CPUInfo does not gather very usefull information on the underlying hardware except the CPU model.
These changes makes the SystemInfo and CPUInfo make precise about the actual board running Kodi.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
The patch has been tested on multiple Amlogic SoCs running Linux 4.18 running the LibreELEC distribution.

The kodi Log now shows:
```
11:11:51.575 T:4098215136  NOTICE: Running on Amlogic Meson GXL (S905X) Libre Technology CC with LibreELEC (community): devel-20180911132354-c6fa5b5 9.1, kernel: Linux ARM 64-bit version 4.18.6
```

## Screenshots (if appropriate):
![Kodi on Libre Computer CC](https://user-images.githubusercontent.com/492099/45432274-c378c400-b6a9-11e8-95d1-d58b14910863.jpg)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

superna9999 committed:
utils: SystemInfo: Use sysfs-devices-soc on Linux if available

New Linux embedded systems uses the new sysfs soc device interface
to store the SoC family, id and revision and the board model.

Use this for the Kodi SystemInfo.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
utils: CPUInfo: use sysfs-devices-soc to gather more system info

New Linux embedded systems uses the new sysfs soc device interface
to store the SoC family, id and the board model.

Use this for the Kodi CPUInfo.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

Rechi commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "CPUInfo.h"
 #include "utils/log.h"
 #include "utils/Temperature.h"
+#include <utils/SysfsUtils.h>
```
change to `#include "utils/SysfsUtils.h"` and move it up one line as  the includes should be sorted (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#6-headers)

Rechi commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -419,6 +420,28 @@ CCPUInfo::CCPUInfo(void)
     }
     fclose(fCPUInfo);
+    // new socs uses the sysfs soc interface to describe the hardware
+    if (SysfsUtils::Has("/sys/bus/soc/devices/soc0")) {
```
curly bracket into new line (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces)

chewitt replied:
Franglais correction: s/uses/use

Rechi commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -419,6 +420,28 @@ CCPUInfo::CCPUInfo(void)
+            SysfsUtils::GetString("/sys/bus/soc/devices/soc0/revision", revision);
+        if (!machine.empty())
+            m_cpuHardware = machine;
+        if (!family.empty() && !soc_id.empty()) {
```
curly bracket into new line (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces)

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -72,6 +72,7 @@ using namespace winrt::Windows::System::Profile;
 #include <sys/param.h>
 #elif defined(TARGET_LINUX)
 #include <linux/version.h>
+#include <utils/SysfsUtils.h>
```
change to `#include "utils/SysfsUtils.h"` and move it into the above block, as Kodi headers should be included before system headers (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#6-headers)

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -764,6 +765,17 @@ std::string CSysInfo::GetManufacturerName(void)
     auto manufacturer = eas.SystemManufacturer();
     g_charsetConverter.wToUTF8(std::wstring(manufacturer.c_str()), manufName);
+#elif defined(TARGET_LINUX)
+    if (SysfsUtils::Has("/sys/bus/soc/devices/soc0/family")) {
```
curly bracket into new line (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces)

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -764,6 +765,17 @@ std::string CSysInfo::GetManufacturerName(void)
+    if (SysfsUtils::Has("/sys/bus/soc/devices/soc0/family")) {
+        std::string family;
+        SysfsUtils::GetString("/sys/bus/soc/devices/soc0/family", family);
+	if (SysfsUtils::Has("/sys/bus/soc/devices/soc0/soc_id")) {
```
curly bracket into new line (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces)

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -764,6 +765,17 @@ std::string CSysInfo::GetManufacturerName(void)
+            std::string soc_id;
+            SysfsUtils::GetString("/sys/bus/soc/devices/soc0/soc_id", soc_id);
+            manufName = family + " " + soc_id;
+        } else
```
curly bracket into new line (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-braces)

Rechi commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -419,6 +420,30 @@ CCPUInfo::CCPUInfo(void)
+    // new socs use the sysfs soc interface to describe the hardware
+    if (SysfsUtils::Has("/sys/bus/soc/devices/soc0"))
+    {
+        std::string machine, family, soc_id, revision;
```
indentation is wrong for the whole PR, https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-indentation

chewitt commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -419,6 +420,30 @@ CCPUInfo::CCPUInfo(void)
+        if (m_cpuModel.empty())
+          m_cpuModel = family + " " + soc_id;
+        else
+          m_cpuModel += "(" + family + " " + soc_id + ")";
```
I tested this earlier and I'd remove the extra brackets here as the /sys/soc_id on Amlogic outputs `GXBB (S905-H)` so the compound GUI entry shows `ARMv8 Processor rev 4 (v8l) (Amlogic Meson GXBB (S905-H))` which is ugly. The (IMHO better) alternative would be to add another line item for `soc: Amlogic Meson GXBB (S905-D)` that's separate from the CPU details.

superna9999 replied:
@chewitt I agree for the extra brackets, for the SoC entry it will need more work, and it could even split into "SoC Family" and "SoC Model"

superna9999 replied:
I added a new "SoC" line and moved the soc family and id to this identifier, I also dropped the revision which is too detailed and not related to the board revision.

superna9999 committed:
utils: Add CPU SoC information

Add a new line in GUIWindowSystemInfo to display the actual
SoC (System-On-Chip) used on the system only if the information
is available, otherwise it's not showed.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

chewitt commented:
@Rechi i'm travelling at the moment (road trip) so it will be another week before I can build/test with the updates Neil pushed.

@MilhouseVH can you also test and see what this looks like on pi hardware please.

MilhouseVH commented:
LibreELEC RPi3+ before:
![s1](https://i.imgur.com/NCum73H.jpg)

LibreELEC RPi3+ after:
![s2](https://i.imgur.com/mUXVMZx.jpg)

LibreELEC x86_64 is unchanged.

wsnipex commented:
Imho, we shouln't display the SoC line if it is empty.

superna9999 commented:
@wsnipex this means the `CoreInfo::m_cpuSoC`variable and `CSysInfo::GetCPUSoC()` would behqve differently from BogoMips/Hardware/Revision/Serial.
This means I should leave `CoreInfo::m_cpuSoC` empty, and not prepend `SoC: `in `CSysInfo::GetCPUSoC()` and then add a conditional `SET_CONTROL_LABEL()`

wsnipex commented:
I'm only talking about the system info screen. Just don't show the line if SoC is N/A

superna9999 commented:
@wsnipex Made the changes, only appears if the SoC info is available
@MilhouseVH can you run your nighly with this updated patch ? You should not see the SoC line anymore

wsnipex reviewed (approved):
looks good to me
